### PR TITLE
ASoC: Intel: sof_sdw: add support for systems without i915 audio

### PR DESCRIPTION
**Pull request by kv2019i**
Extend the generic SOF Soundwire machine driver to support systems where
iDisp HDMI/DP audio codec is disabled for some reason (i915 driver
disabled, HDMI/DP implemented with a discrete GPU, etc). Switch codecs
to SoC dummy in the affected DAI links. This allows to reuse existing
topologies for this case.

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -704,11 +704,14 @@ static inline int get_next_be_id(struct snd_soc_dai_link *links,
 	return links[be_id - 1].id + 1;
 }
 
+#define IDISP_CODEC_MASK	0x4
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw.c:707]: @kv2019i  One question: is our HDMI codec fixed at 0x4 ? thanks!

Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c:707]: @RanderWang Good question. This is basicly defining a bit in HDA.STATESTS register which maps to SDI interface. On all hw we have, two first SDI signals are reserved for external HDA codecs while 3rd signal is for iDisp. In practise assumption on 0x4 is used throughout the codebase now. I tried to align the code with generic HDA driver, using same names and expressoins, so it's easier to see the similarity.

**Comment by kv2019i** [comment]
@plbossart wrote:
> Looks good to me @kv2019i but out of curiosity was this tested on a real device or fork-lifted based on the skl-generic-hda card stuff?

Yes, I did test this on a TGL platform with SDW codecs (and used a test patch to inject an error to i915 probe).

This patch also allows to disable HDMI/DP audio at runtime using module parameter:
options snd_sof_intel_hda_common codec_mask=0x0

Tested this as well.

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sof_sdw.c:707]: thanks ! I finally get the answer.

File: sound/soc/intel/boards/sof_sdw_hdmi.c [hunk 2]
```diff
@@ -52,6 +52,12 @@ int sof_sdw_hdmi_card_late_probe(struct snd_soc_card *card)
 	int err, i = 0;
 	char jack_name[NAME_SIZE];
 
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw_hdmi.c]: Under which conditions can this happen? Is this really related to this patch? Does this patch really add a possibility for this function to be called with this list empty? And even if this does now become possible: shouldn't this be `ENODEV`?

Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw_hdmi.c]: @lyakh Fair point that this should perhaps be in a separate patch. This same check is in other machine drivers (and EINVAL is used in those) and protects against the case where topology does not have PCMs for HDMI. In this case, driver should be fail gracefully and not crash. Let me split this out into a separate patch.

Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw_hdmi.c]: @lyakh Now moved to a separate branch, please check.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -906,13 +912,18 @@ static int sof_card_dai_links_create(struct device *dev,
...
+								      i + 1);
+			if (!idisp_components[i].dai_name)
+				return -ENOMEM;
+		} else {
+			idisp_components[i].name = "snd-soc-dummy";
+			idisp_components[i].dai_name = "snd-soc-dummy-dai";
+		}
 
 		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
 					  "iDisp%d Pin", i + 1);
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw.c:929]: @kv2019i   I tested it on TGL.  If HDMI_CODEC == NULL,  I will get error when doing playback with HDMI. you need to change cpu_name to another one instead of iDisp Pin.
````
[   46.911750] sof-audio-pci 0000:00:1f.3: ASoC: error at snd_soc_dai_hw_params on iDisp1 Pin: -22
[   46.911754]  iDisp1: ASoC: hw_params BE failed -22
[   46.911755]  HDMI1: ASoC: hw_params BE failed -22
````

Comment (RanderWang) [reply sound/soc/intel/boards/sof_sdw.c:929]: I mean ctx->idisp_codec == false

Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c:929]: @RanderWang That is expected failure -- there's no HDMI/DP codec attached, so error is expected. When ctx->idisp_codec is false, the HDMI controls will not be created and we have UCM rules to check this. So e.g. Pulseaudio will ignore the HDMI PCM nodes in this case. So normal applications should never hit this error. If we'd use a different cpu_name, then topology load would fail and this is specifically what we want to avoid.

**Comment by kv2019i** [comment]
@lyakh wrote:
> just to make sure we realise, that we're relying on kzalloc() to set idisp_codec to false...

Ack, this is intentional. The kzalloc is in the same fine, so should be obvious enough (or so I thought).

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sof_sdw.c:929]: thanks,  I get it.
